File: packages/ns-wry/platforms/ios/NSCWry/Sources/NSCWry/NSCWry.swift
import Foundation

/// NSCWryRuntime is the primary entry-point for the wry engine on iOS.
/// It is exposed to NativeScript via `@objc(NSCWryRuntime)`.
@objc(NSCWryRuntime)
open class NSCWryRuntime: NSObject {

    private let stackSize: UInt32
    private var disposed = false

    @objc(initWithStackSize:)
    public init(stackSizeInBytes: UInt32) {
        self.stackSize = stackSizeInBytes
        super.init()
    }

    @objc(wryVersion)
    public static func wryVersion() -> String {
        return "0.1.0"
    }

    @objc(initRuntime)
    open func initRuntime() {
        // Stub: a real implementation would initialize the wry engine.
    }

    @objc(eval:)
    open func eval(_ script: String) -> String {
        guard !disposed else { return "" }
        // Stub: a real implementation would evaluate JS in the WebView.
        return ""
    }

    @objc(loadUrl:)
    open func loadUrl(_ url: String) {
        guard !disposed else { return }
    }

    @objc(setHtml:)
    open func setHtml(_ html: String) {
        guard !disposed else { return }
    }

    @objc(isLoaded)
    open func isLoaded() -> Bool {
        return !disposed
    }

    @objc(callWithArgs:)
    open func call(withArgs args: [Any]) -> Any? {
        _ = args
        return nil
    }

    @objc(dispose)
    open func dispose() {
        disposed = true
    }
}

File: packages/ns-wry/platforms/android/nsc-wry-android/library/src/main/kotlin/org/nativescript/wry/NSCWry.kt
package org.nativescript.wry

class NSCWryException(message: String) : RuntimeException(message)

// NSCWry — Kotlin wrapper around the wry native library (libwry_jni.so),
// consumed by the NativeScript Android runtime.

class NSCWryRuntime @JvmOverloads constructor(
    @Suppress("UnusedPrivateProperty") stackSizeInBytes: Int = 65536,
) : AutoCloseable {
    internal var handle: Long = 0
        private set
    private var closed = false

    companion object {
        @JvmStatic
        fun wryVersion(): String = NativeWry.version()
    }

    init {
        handle = 1L // non-zero handle = created
        val err = NativeWry.init()
        if (err != 0) throw NSCWryException("failed to initialize wry runtime")
    }

    fun eval(script: String): String {
        checkNotClosed()
        return NativeWry.eval(handle, script)
    }

    fun loadUrl(url: String) {
        checkNotClosed()
        val err = NativeWry.loadUrl(handle, url)
        if (err != 0) throw NSCWryException("failed to load URL: $url")
    }

    fun setHtml(html: String) {
        checkNotClosed()
        val err = NativeWry.setHtml(handle, html)
        if (err != 0) throw NSCWryException("failed to set HTML content")
    }

    fun isLoaded(): Boolean = !closed

    fun call(
        @Suppress("UnusedParameter") withArgs: Array<Any>,
    ): Any? {
        checkNotClosed()
        // Stub: a real implementation would dispatch a call to the WebView.
        return null
    }

    override fun close() {
        if (closed) return
        closed = true
        if (handle != 0L) {
            NativeWry.dispose(handle)
            handle = 0
        }
    }

    private fun checkNotClosed() {
        if (closed) throw NSCWryException("runtime is closed")
    }
}
